### ISSUE TITLE
GUI freezing/crashing on MacOS Sonoma 14.0

### ISSUE THREAD

**Issue by thumpclang** [pure-data/pure-data#2105]
I just migrated from a 2016 Intel Macbook Pro to a new M2 Macbook Pro, and after (unwisely) upgrading to the latest macOS 14.0 Sonoma before some performances (yes, I’m that dumb...) it seems Pd is having GUI problems.

The M2 machine shipped with macOS Ventura (13.6 if I recall); I briefly tested my performance patches with two of the newest versions of Pd (0.54, 0.53-2), both in Rosetta 2 and natively in Apple Silicon, during which I didn’t notice any unusual behavior. I also loaded a GUI-intensive patch in Pd 0.49-1 (my massive GUI is significantly more performant on MacOS in 0.49-1 versus newer Pd versions, I’m guessing for Tcl/Tk reasons). That also seemed to work, although I didn’t have time for more than a quick test.

After upgrading to Sonoma, all 3 Pd versions are having issues. 0.54 and 0.53-2 (with & without Rosetta) will intermittently stop responding to mouse clicks, both in the patch window and in menu dialogs (e.g., Audio Settings), sometimes freezing for 30+ seconds before resuming normal interaction. 0.49-1 will load, but crashes immediately when mousing over the top menu when “menu hiding” is enabled, which was not an issue in Ventura. If I disable menu hiding, 0.49-1 seems stable & doesn’t suffer from the mouse click issues of the newer Pd versions.

For my performances I ended up running the audio core of my patch in 0.54 from the command line (-nogui), and ran the GUI in Purr Data, communicating between the two over netsend. This worked fine, which perhaps suggests that the underlying issue is with Tcl/Tk or pd-gui within Sonoma?

I don’t know how involved a fix for this would be in light of the fact that Apple & Tcl/Tk don’t seem to play nice with each other in general these days (see #1488), but I mainly wanted to report this as a cautionary tale for other Mac users who might get bitten by this as well.

**Comment by Ant1r** [comment]
I experimented this mouse click bug too on Sonoma. That seems to affect all TclTk (or Tkinter) apps on this OS.
Holding down the mouse button and moving slightly the pointer can help as a temporary workaround.

Here are some references to this bug I could find:
https://github.com/Anjok07/ultimatevocalremovergui/issues/840
https://www.reddit.com/r/MacOS/comments/16tgy8e/intermittent_mouse_click_issues_after_sonoma/?rdt=34116
https://discussions.apple.com/thread/255171864

**Comment by Ant1r** [comment]
From https://github.com/Anjok07/ultimatevocalremovergui/issues/840#issuecomment-1751688579 , we can read:

> How did you solve the issue?

> I  updated tk/tcl from 8.6.12 to 8.6.13

so that seems to be the way to go...

**Comment by porres** [comment]
> so that seems to be the way to go...

confirmed here

https://github.com/sebshader/pdnext/issues/25#issuecomment-1768503129

**Comment by danomatika** [comment]
Thanks for finding the solution to this already. I have not upgraded any of my machines (home or work) to macOS 14 as I am in the middle of project/development cycles.

I will try to look into this tonight which mainly involves double-checking if we need to continue making and applying our custom patches, mainly to fix key handling. I can also make a test build for people to try.

**Comment by danomatika** [comment]
Also, in the future, please include me on these kinds major of macOS-related issues by using @danomatika. I have notifications turned off on this repo so I don't see everything that's going on everyday (I would go crazy).

**Comment by danomatika** [comment]
Ok, here is an arm64 build of Pd 0.54-0 from the latest `develop` branch with Tcl/Tk 8.6.13:

http://docs.danomatika.com/pdbuilds/0.54/Pd-0.54-0-arm64-tk8613.zip


To try at home, ssuming JACK universal is installed https://jackaudio.org/downloads/:

~~~
mac/tcltk-wish.sh 8.6.13
./configure --enable-jack --disable-jack-framework
make
mac/osx-app.sh -w Wish-8.6.13.app 0.54-0
mv Pd-0.54-0.app Pd-0.54-0-arm64-tk8613.app
~~~

_Note: There is still a problem building a universal Wish.app, unfortunately. I haven't looked into this yet._

**Comment by danomatika** [comment]
I have a added a separate build for x86_64 (Intel) and updated the previous post with info on building for a single specific arch.

**Comment by porres** [comment]
should the intel one work in an old 10.14.6 Mojave system? Cause...

![Screen Shot 2023-10-22 at 20 41 09](https://github.com/pure-data/pure-data/assets/5360142/94f4c9ae-2f4c-491f-ad8c-595d56b1e03b)

**Comment by danomatika** [comment]
Would be nice to see more of what's there starting with Exception Type. Maybe scroll down and post second screenshot, don't need copy/paste of everything.

**Comment by porres** [comment]
I was able to capture it all in here

<img width="998" alt="Screen Shot 2023-10-22 at 20 49 16" src="https://github.com/pure-data/pure-data/assets/5360142/2acb8f71-9d4b-4111-b026-49b2742926ec">
<img width="994" alt="Screen Shot 2023-10-22 at 20 49 31" src="https://github.com/pure-data/pure-data/assets/5360142/22798e11-52b7-480f-9dcc-6ba8d5337bd0">

**Comment by porres** [comment]
tested on 12.3.1, monterey, and it works!

**Comment by porres** [comment]
the intel one, no aple silicon macs pin my household yet :)

**Comment by danomatika** [comment]
It looks like TK's Wish now weak links to `/System/Library/Frameworks/UniformTypeIdentifiers.framework` which is only available on macOS 11+. It should be possible to build Wish without it, but I can't do that tonight.

**Comment by thumpclang** [comment]
> Ok, here are arm64 (Apple Silicon) and x86_64 (Intel) builds of Pd 0.54-0 from the latest `develop` branch with Tcl/Tk 8.6.13:

Both tested on Sonoma, no mouse click problems or crashing! Thank you, @danomatika

> There is still a problem building a universal Wish.app with tcltk-wish.sh, unfortunately. I haven't looked into this yet. Building for a single arch does work.

Aha! I was wondering why my attempt to build a universal 8.6.13 Wish.app failed.

> 0.49-1 will load, but crashes immediately when mousing over the top menu when “menu hiding” is enabled, which was not an issue in Ventura.

This issue persists, and although disabling menu hiding seems to remedy the immediate crashing, it makes me a bit nervous to use 0.49-1 in performance in case there are other undiscovered gremlins lurking. More generally, is there a sense of whether it's worth trying to solve the sluggish GUI performance issues with newer Tcl/Tk versions on Mac, given the long term goal of a toolkit-agnostic Pd? (I could also take this question to #1488 if that seems more appropriate)

**Comment by danomatika** [comment]
Honestly, these kinds of issues are part & parcel for cross-platform projects. Something always breaks here and there especially when you have to support so many different system versions. It's frustrating but unless everyone builds and systems on beta release OS versions, you can't catch everything very early.

The real solution is to provide a bug report to the TK devs *with* (this is important) a simple Tcl script which reproduces the problem, otherwise someone has to dive into the Tk source code and "figger it out." I have done the latter and have very little desire to have to do that again however the patches we provided to the Tk devs were incorporated and key handling was fixed for all. Yeah.

**Comment by thumpclang** [comment]
Totally understood!

**Comment by sebshader** [comment]
not sure this is the place to discuss this, but I had a discord user compile universal 8.6.13 for me and it seems like it was successful. they just used ```./tcltk-wish.sh -l --universal 8.6.13```. A compiled version is https://github.com/sebshader/pdnext/blob/4d1698113530a04ea405daa821f573bbd5cd300c/mac/stuff/wish-shell.tgz

**Comment by danomatika** [comment]
> not sure this is the place to discuss this, but I had a discord user compile universal 8.6.13 for me on Ventura and it seems like it was successful. they just used `./tcltk-wish.sh -l --universal 8.6.13`. A compiled version is https://github.com/sebshader/pdnext/blob/4d1698113530a04ea405daa821f573bbd5cd300c/mac/stuff/wish-shell.tgz

Can you get info about their system? I'm still on macOS 13 and Xcode 14. Maybe this is fixed for 14 or Xcode 15...

**Comment by danomatika** [comment]
Ok, I have added a universal build as well:

http://docs.danomatika.com/pdbuilds/0.54/

To try at home, using the universal Wish renamed to `Wish-8.6.13-universal.app`:
~~~
./configure --enable-jack --disable-jack-framework --enable-universal="arm64,x86_64"
make
mac/osx-app.sh -w Wish-8.6.13-universal.app 0.54-0
mv Pd-0.54-0.app Pd-0.54-0-universal-tk8613.app
~~~

**Comment by umlaeute** [comment]
@danomatika where's the "universal Wish" from? is it https://github.com/sebshader/pdnext/blob/4d1698113530a04ea405daa821f573bbd5cd300c/mac/stuff/wish-shell.tgz?

**Comment by danomatika** [comment]
Yes.enohp ym morf tnes-----------Dan Wilcoxdanomatika.comrobotcowboy.comOn Oct 23, 2023, at 11:44 AM, umläute ***@***.***> wrote:﻿
@danomatika where's the "universal Wish" from? is it https://github.com/sebshader/pdnext/blob/4d1698113530a04ea405daa821f573bbd5cd300c/mac/stuff/wish-shell.tgz?

**Comment by sebshader** [comment]
not sure but it was probably xcode 14 since they insntalled it in january 2023 [@bvvt] https://github.com/bvvt

**Comment by danomatika** [comment]
I've backported a [fix for the builds not working on pre macOS 11 systems](https://github.com/tcltk/tk/commit/a3eb1edc6e9c2eb227b40c7d46f42e6162be78d6) as a patch to Tk 8.6.13 in the `develop` branch. New test builds marked "-fixed" should work on both new and older systems, as before:

http://docs.danomatika.com/pdbuilds/0.54/

It also might be good for someone to try the current bleeding edge Tcl/Tk 8.6 development branch as there is lots of activity right now. I believe we may see 8.6.14 next month, like 8.6.13 last year. To do this using the Wish build script:
~~~
tcltk-wish.sh --git 8.6.13-git -b core-8-6-branch
~~~

**Comment by thumpclang** [comment]
> It also might be good for someone to try the current bleeding edge Tcl/Tk 8.6 development branch as there is lots of activity right now.

I just tried this, and now have a non-universal Pd 0.54 (from git) with Wish-8.6.13-git. All seems fine with a quick test (opening help patches, clicking around, confirming audio output). Is there anything else in particular that one should do to stress test a build?

FWIW, the sluggish GUI performance seems no different using this git version of 8.6.13.

**Comment by porres** [comment]
> I've backported a fix for the builds not working on pre macOS 11 systems from [this TK commit](https://github.com/tcltk/tk/commit/a3eb1edc6e9c2eb227b40c7d46f42e6162be78d6) as a patch to Tk 8.6.13 in the develop branch. New test builds marked "-fixed" should work on both new and older systems, as before:

not quite fixed for me yet :/

![Screen Shot 2023-10-23 at 21 36 12](https://github.com/pure-data/pure-data/assets/5360142/3a69f28e-762a-4c4f-a68e-ad46ea8e12e9)

![Screen Shot 2023-10-23 at 21 36 24](https://github.com/pure-data/pure-data/assets/5360142/a75c31ea-7bb2-44c4-ad45-ca746afa6c16)


![Screen Shot 2023-10-23 at 21 36 33](https://github.com/pure-data/pure-data/assets/5360142/2a58a3ba-ee5b-45f4-a50e-53aa47dc9932)

**Comment by thumpclang** [comment]
From @umlaeute on PD-Dev:

> The binaries found here
> <https://git.iem.at/pd/pure-data/-/jobs/57601/artifacts/raw/Pd-0.54-0-61-g3afa2ff6.dmg>
> should be build with a Tcl/Tk-8.6.13 that runs on anything since OSX-10.9 Mavericks, either x86_64 and arm64.
> 
> i cannot test right now, so it would be great if people could do that while i sleep. i'm esp interested in the extremes, that is:
> - old macOS versions & old OSX versions (Mountain Lion anybody? but really any version that is no longer supported by Apple would be great)
> - newest macOS versions (i can do some basic tests on Monterey tomorrow, but don't have access to Ventura or Sonoma machines)

How's this for extremes?: this build works for me on:
-2023 14" M2 Macbook Pro (Sonoma 14.0)
-2008 Core 2 Duo Macbook (Lion 10.7.5)

From @danomatika also on PD-Dev:

> Ok, this back ported patch is in the latest develop branch commit. I have made new test builds with Tk 8.6.13 + the patch applied so the framework is now weakly linked and these should work on older macOS versions, as before.
> 
> New builds are marked "-fixed" and found at:
> 
> http://docs.danomatika.com/pdbuilds/0.54/

Pd-0.54-0-universal-tk8613 seems to work fine on Sonoma.

I also tried running the Pd that's zipped in Pd-0.54-0-x86_64-tk8613-fixed.zip on the ancient Lion Macbook, no dice there (although I don't know if you were expecting this one to run on such old OSes). Here's a console crash log if it's relevant:
![Pd-0 54-0-x86_64-tk8613_Lion](https://github.com/pure-data/pure-data/assets/88018389/f1ac7aeb-68fa-4712-a7c2-1d1e5836088b)

**Comment by umlaeute** [comment]
@thumpclang yes that would be the extremes i'm interested in.
could you please try the binary from https://git.iem.at/pd/pure-data/-/jobs/57601/artifacts/raw/Pd-0.54-0-61-g3afa2ff6.dmg (it's different from the one that dan posted).

**Comment by umlaeute** [comment]
now that the CI-artifacts are downloadable again, i checked on Monterey (arm64, amd64) and Mavericks (amd64), and both start up.
jack2 doesn't work on Mavericks (and even crashes when selected), but that seems to be a problem with jack itself. (e.g. qJackCtl also crashes), and is not something i would like to spend time on.

apart from that, i cannot really test (these are remote and/or virtualized machines), but so far this looks good.

this is using a universal Wish from unpatched Tcl/Tk sources, but with the amd64 portion being compiled on Mavericks and then using `lipo` to merge the amd4 and arm64 binaries.
this will obviously not enable any features that are only available on macOS>>10.9.
The most apparent difference is, that the Wish will not support dark mode (or sth like this; it stays "light", whereas Wish compiled on Sierra or newer shows up as "dark" (both tested on Monterey).

i don't know what the `UniformTypeIdentifiers` framework is supposed to do.

@danomatika would you expect any problems with such a build?

**Comment by thumpclang** [comment]
> could you please try the binary from https://git.iem.at/pd/pure-data/-/jobs/57601/artifacts/raw/Pd-0.54-0-61-g3afa2ff6.dmg (it's different from the one that dan posted).

@umlaeute Unless that link changed between when you posted it on PD-Dev & your more recent comment, that’s the build I was referring to when I said it worked for me on both Sonoma & Lion. (sorry if that was confusing formatting, I lumped my test results for both your build & Dan’s builds into one comment…)

I haven’t been testing Jack functionality as I don’t have it installed on either Mac, could attempt to install it & test with these builds tomorrow if that’s a particular point of interest.

**Comment by umlaeute** [comment]
@thumpclang no the link did not change.
 i just missed the line that said that "this build works for me on" (and instead thought that you have these two machines and then i read on that dan's builds where failing on Lion).

so sorry for the noise.


@danomatika if you are ok with that i would then use my universal builds for the official Pd releases.

**Comment by umlaeute** [comment]
i forgot maybe @porres could do a test with https://git.iem.at/pd/pure-data/-/jobs/57601/artifacts/raw/Pd-0.54-0-61-g3afa2ff6.dmg  as well

**Comment by danomatika** [comment]
> @danomatika if you are ok with that i would then use my universal builds for the official Pd releases.
> 
> **EDIT**: i forgot maybe @porres could do a test with https://git.iem.at/pd/pure-data/-/jobs/57601/artifacts/raw/Pd-0.54-0-61-g3afa2ff6.dmg as well

Wait. I made a mistake and the "fixed" x86_64 build does not weak link, hence @porres crash. I would much prefer *not* to have to rely on building on older system versions, if at all possible.

**Comment by umlaeute** [comment]
fine with me (if it works). the baseline macOS we should support is macOS 10.10 Yosemite 

could you do a test build?

**Comment by danomatika** [comment]
@porres @thumpclang Can y'all test the "x86_64 fixed2" build from:

http://docs.danomatika.com/pdbuilds/0.54/

The previous "x86_64 fixed" build was *not* weak linked. My mistake. I am juggling quite a few Wish versions right now.

**Comment by umlaeute** [comment]
i did a test run of the `x86_64 fixed2` build on Mavericks and it opens up!

**Comment by danomatika** [comment]
Ok, cool. The only thing needed for this issue was the patch I've added to the `develop` branch.

It might be worth checking if there are another other things worth back porting such as PDF printing (instead of PS), but I would do that in a feature branch.

**Comment by thumpclang** [comment]
> Can y'all test the "x86_64 fixed2" build from:
http://docs.danomatika.com/pdbuilds/0.54/

`x86_64 fixed2` is running fine in Sonoma. It crashes in Lion 10.7.5, but maybe that's to be expected if Yosemite is the oldest OS you're aiming for? I have a Lion crash log if you'd like to see for whatever reason.

Also woke up to see the Pd-0.54-1test1.1 build from Miller, seems to work well in Sonoma! (running either Apple Silicon or Rosetta 2)

@danomatika @umlaeute  Thanks so much for your work on this!

**Comment by esaruoho** [comment]
i can confirm that i am now fighting with this same issue. with 0.53-2 and 0.54-0, it takes a strong (hard) click to get the context menu to open, and i cannot simply "click" on an abstraction to open it. it is quite harrowing. is there a build i can try that might fix this? PD is basically unusable for me re: gigging purposes now.

**Comment by esaruoho** [comment]
> Can y'all test the "x86_64 fixed2" build from:
> 
> http://docs.danomatika.com/pdbuilds/0.54/
> 
> The previous "x86_64 fixed" build was _not_ weak linked. My mistake. I am juggling quite a few Wish versions right now.

amazing! tk8613-fixed2 fixed the mouse-clicking issues! much appreciated!!
@danomatika

**Comment by umlaeute** [comment]
for now, please stop testing the builds from @danomatika and instead focus on [Pd-0.54-1test1](https://msp.ucsd.edu/Software/pd-0.54-1test1.macos.zip) from https://msp.ucsd.edu/software.html

**Comment by umlaeute** [comment]
@thumpclang  and yes, Yosemite is the oldest release we are aiming for (but in reality I do not have a Yosemite machine, only a Mavericks one).
JACK is known to crash on that platform, but otherwise it should work (i don't know to which extent)

**Comment by sebshader** [comment]
can the mac/stuff/wish-shell.tgz be updated as well?

**Comment by umlaeute** [comment]
We typically do not do that, as mac/stuff/wish-shell.tgz is required to build Pd for very old OSX versions.

**Comment by porres** [comment]
Pd-0.54-1test1 works fine

**Comment by porres** [comment]
so, if the release solves this issue, we can close it :)

**Comment by thumpclang** [comment]
For good measure: I just updated Sonoma to 14.1 today to see if Apple's latest bug fixes made any difference.

0.54-1test1 seems to be working fine! 🎉

Also tried the release version of 0.54-0 out of curiosity, that one is still having the same mouse click issues & some other GUI weirdness. So it seems whatever work Apple did under the hood between 14.0 and 14.1 didn't make a difference for Tcl/Tk 8.6.12

And, as per my [original comment](https://github.com/pure-data/pure-data/issues/2105#issue-1938357286) in this thread, 0.49-1 still crashes immediately when revealing the top menu bar when "menu hiding" is enabled. The workaround I found before (disabling menu hiding) still allows for 0.49-1 to run normally in Sonoma 14.1.
